core/controllers/topic_editor_test.py

```diff
@@ -0,0 +1,182 @@
+        story_id = json_response['storyId']
+        self.assertEqual(len(story_id), 12)
+        self.assertTrue(
+            story_services.get_story_by_id(story_id, strict=False) is not None)
+        self.logout()
+
+
+class TopicEditorTest(BaseTopicEditorControllerTest):
+
+    def test_assign_topic_manager_role(self):
+        """Test the assign topic manager role for a topic functionality.
+        """
+        self.login(self.ADMIN_EMAIL, is_super_admin=True)
+        with self.swap(feconf, 'ENABLE_TOPIC_PAGE', True):
+            response = self.testapp.get(
+                '%s/%s' % (feconf.TOPIC_EDITOR_URL_PREFIX, self.TOPIC_ID))
+            csrf_token = self.get_csrf_token_from_response(response)
+
+        # Test for when assignee does not have sufficient rights to become a
+        # manager for a topic.
+        json_response = self.put_json(
+            '%s/%s/%s' % (
+                feconf.TOPIC_MANAGER_PREFIX, self.TOPIC_ID, self.new_user_id),
+            {}, csrf_token=csrf_token, expect_errors=True,
+            expected_status_int=401)
+        self.assertEqual(json_response['status_code'], 401)
+
+        # Test for valid case.
+        json_response = self.put_json(
+            '%s/%s/%s' % (
+                feconf.TOPIC_MANAGER_PREFIX, self.TOPIC_ID,
+                self.topic_manager_id),
+            {}, csrf_token=csrf_token, expect_errors=True,
+            expected_status_int=200)
+        self.assertEqual(json_response['role_updated'], True)
+        self.logout()
+
+        # Test for when committer doesn't have sufficient rights to assign
+        # someone as manager.
+        json_response = self.put_json(
+            '%s/%s/%s' % (
+                feconf.TOPIC_MANAGER_PREFIX, self.TOPIC_ID, self.new_user_id),
+            {}, csrf_token=csrf_token, expect_errors=True,
+            expected_status_int=401)
+        self.assertEqual(json_response['status_code'], 401)
+
+    def test_access_topic_editor_page(self):
+        """Test access to editor pages for the sample topic."""
+
+        # Check that non-admin and topic_manager cannot access the editor page.
+        response = self.testapp.get(
+            '%s/%s' % (feconf.TOPIC_EDITOR_URL_PREFIX, self.TOPIC_ID))
+        self.assertEqual(response.status_int, 302)
+
+        # Check that topic admins can access and edit in the editor
+        # page.
+        self.login(self.ADMIN_EMAIL)
+        with self.swap(feconf, 'ENABLE_TOPIC_PAGE', True):
+            response = self.testapp.get(
+                '%s/%s' % (feconf.TOPIC_EDITOR_URL_PREFIX, self.TOPIC_ID))
+            self.assertEqual(response.status_int, 200)
+        self.logout()
+
+    def test_editable_topic_handler_get(self):
+        # Check that non-admins cannot access the editable topic data.
+        response = self.testapp.get(
+            '%s/%s' % (feconf.EDITABLE_TOPIC_DATA_URL_PREFIX, self.TOPIC_ID))
+        self.assertEqual(response.status_int, 302)
```

Comment:
Yeah, I thought so too, but even in collection_editor, when a non collection_editor tries to access the page, the status returned is 302. But, if the decorator is can_access_admin_page, the status returned is 401. Not sure why, even though the same exception is raised in acl_decorators.py

Reply:
I think that's a bug that might need debugging...

Reply:
I think it might depend on the request. For tests related to TOPIC_EDITOR_DATA_URL_PREFIX, put_json returns 401 error, while testapp.get() and testapp.delete() returns 302.